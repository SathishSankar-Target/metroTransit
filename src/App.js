import React , { Component }from 'react';
import './App.css';
import {fetchApi} from './apiCalls'
import DepartureList from './departureList'

class App extends Component {
  constructor(props) {
    super(props)
    this.updateDepartureData = this.updateDepartureData.bind(this)
    this.state = {
      routes : [],
      selectedRoute: '',
      direction: [],
      selectedDirection: '',
      stop: [],
      selectedStop: '',
      showDirection: false,
      showStop: false,
      departureList: [],
      showDepartureList: false,
      showRouteSelection: true,
      showStopSelection: false
    }
  }

  componentDidMount() {
    fetchApi('routes').then(routes => this.setState({routes :routes }))
    setInterval(this.updateDepartureData, 60000)
  }
  // Makes the call every 1 min only when there is selectedRoute, selectedDirection, selectedStop
  async updateDepartureData() {
    if(this.state.selectedRoute !== '' && this.state.selectedDirection !== '' && this.state.selectedStop !== '') {
      const query = this.state.selectedRoute+'/'+this.state.selectedDirection+'/'+this.state.selectedStop
      fetchApi(query).then(departureList => this.setState({departureList: departureList}))
    }
  }

  // Makes the call to get the direction select box data and set/change some default state values
  handleRoutChange = (event) => {
    let value = event.target.value
    this.setState({
      showDirection: false,
      direction: [],
      selectedRoute:'',
      showStop: false,
      departureList: [],
      showDepartureList: false,
    })
    if (value !== 'Select route') {
      fetchApi('directions/'+value)
      .then(direction => this.setState({direction :direction, showDirection: true, selectedRoute: value }))
    } 
  }

  // Makes the call to get the stop select box data and set/change some default state values
  handleDirectionChange = (event) => {
    let value = event.target.value
    this.setState({
      showStop: false,
      stop: [],
      selectedDirection: '',
      departureList: [],
      showDepartureList: false,
    })
    if (value !== 'Select direction') {
      fetchApi('stops/'+this.state.selectedRoute+'/'+value)
      .then(direction => this.setState({stop :direction, showStop: true, selectedDirection: value }))
    } 
  }

  // Makes the call to get the list departure data and set/change some default state values
  handleStopChange = (event) => {
    let value = event.target.value
    this.setState({
      showDepartureList: false,
      departureList: []
    })
    if (value !== 'Select stop') {
      fetchApi(this.state.selectedRoute+'/'+this.state.selectedDirection+'/'+value)
      .then(departureList => this.setState({departureList: departureList, showDepartureList: true, selectedStop: value}))
    } 
  }

  render() {
    const state = this.state
    const routeData = state.routes 
    const directionData = state.direction
    const stopData = state.stop
    const departureList = state.departureList
    // const { routeData, directionData, stopData, departureList } = state
    return (
      <div className="App">
        <h2 className="page-title">Real-time Departures</h2>
        <div className="select-box-container">
        {/* Select the route select box */}
        { routeData.length ? 
          <select className="select-box" onChange={(e) => this.handleRoutChange(e)}>
            <option>Select route</option>
            { routeData.map(route => <option key={route.RouteId} value={route.RouteId}>{route.Description}</option>)}
          </select> :
          (routeData.errors && <div>We are unable to get the desired route option for you to select, Kindly try again after some time</div>) }
        {/* Select the direction select box */}
        { (state.showDirection && directionData) &&
          <select className="select-box" onChange={(e) => this.handleDirectionChange(e)}>
            <option>Select direction</option>
            { directionData.map(direction => <option key={direction.DirectionId} value={direction.DirectionId}>{direction.DirectionName}</option>)}
          </select>
        }
        {/* Select the stop select box */}
        { (state.showStop && stopData) &&
          <select className="select-box" onChange={(e) => this.handleStopChange(e)}>
            <option>Select stop</option>
            { stopData.map(stop => <option key={stop.PlaceCode} value={stop.PlaceCode}>{stop.Description}</option>)}
          </select>
        }
        </div>
        {/* Departure list along with the stop description and stop id */}
        { state.showDepartureList && <DepartureList departureList = {departureList} /> }
      </div>
    );
  }
}

export default App;
