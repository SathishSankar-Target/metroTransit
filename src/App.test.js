import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import App from './App';
import DepatureList from './depatureList'
import fetchApi from './apiCalls'


afterEach(cleanup);

it("matches snapshot", () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});


test('render the page title', () => {
  const { getByText } = render(<App />);
  const titleElement = getByText(/Real-time Departures/i);
  expect(titleElement).toBeInTheDocument();
});

const departureList = {
  "Stop": {
      "StopId": 17897,
      "Latitude": 44.972836,
      "Longitude": -93.262431,
      "Description": "7th - Park Station"
  },
  "Departures": [
    {
      "Actual": true,
      "BlockNumber": 1443,
      "DepartureText": "6 Min",
      "DepartureTime": "2020-01-04T16:10:00",
      "Description": " Brklyn Ctr Tc / Rapid",
      "Gate": "",
      "RouteId": "C Line",
      "DirectionId": 4,
      "DirectionText": "NORTHBOUND",
      "Terminal": "",
      "Latitude": 44.97129,
      "Longitude": -93.26141
    },
    {
        "Actual": true,
        "BlockNumber": 1438,
        "DepartureText": "16 Min",
        "DepartureTime": "2020-01-04T16:20:00",
        "Description": " Brklyn Ctr Tc / Rapid",
        "Gate": "",
        "RouteId": "C Line",
        "DirectionId": 4,
        "DirectionText": "NORTHBOUND",
        "Terminal": "",
        "Latitude": 44.97777,
        "Longitude": -93.27639
    },
    {
        "Actual": false,
        "BlockNumber": 1441,
        "DepartureText": "4:30",
        "DepartureTime": "2020-01-04T16:30:00",
        "Description": " Brklyn Ctr Tc / Rapid",
        "Gate": "",
        "RouteId": "C Line",
        "DirectionId": 4,
        "DirectionText": "NORTHBOUND",
        "Terminal": "",
        "Latitude": 44.98426,
        "Longitude": -93.29731
    },
    {
        "Actual": false,
        "BlockNumber": 1444,
        "DepartureText": "4:40",
        "DepartureTime": "2020-01-04T16:40:00",
        "Description": " Brklyn Ctr Tc / Rapid",
        "Gate": "",
        "RouteId": "C Line",
        "DirectionId": 4,
        "DirectionText": "NORTHBOUND",
        "Terminal": "",
        "Latitude": 45.01653,
        "Longitude": -93.30832
    }
    ]
  }

  test(`has rendered a the items passed correctly`, () => {
    const { getByRole } = render(<DepatureList departureList = {departureList} />)
    let listItems = getByRole('list')
    expect(listItems.children.length).toEqual(3)
  })

  test('show more/less button when the list length is more than 3', () => {
    const { getByText } = render(<DepatureList departureList = {departureList} />);
    const showMore = getByText(/Show more depature time/i);
    expect(showMore).toBeInTheDocument();
  });

  test('Check the stop description', () => {
    const { getByText } = render(<DepatureList departureList = {departureList} />);
    const stopDescription = getByText(/7th - Park Station/i);
    expect(stopDescription).toBeInTheDocument();
  });

  test('Check the stop number', () => {
    const { getByText } = render(<DepatureList departureList = {departureList} />);
    const stopNumber = getByText(/Stop 17897/i);
    expect(stopNumber).toBeInTheDocument();
  });

  test('Check the button text change when clickeds on show more depture time', () => {
    const { getByText } = render(<DepatureList departureList = {departureList} />);
    fireEvent.click(getByText(/Show more depature time/i));
    const showLess = getByText(/Show less depature time/i);
    expect(showLess).toBeInTheDocument();
  });

  test('Check increase depature list length after clicking on show more depature time', () => {
    const { getByText, getByRole } = render(<DepatureList departureList = {departureList} />);
    fireEvent.click(getByText(/Show more depature time/i));
    let listItems = getByRole('list')
    expect(listItems.children.length).toEqual(4)
  });

  it('Check on intial load has a Select route select box rendred', () => {
    const { getByText } = render(<App />);
    expect(getByText(/Select route/i)).toBeInTheDocument();
  });


  