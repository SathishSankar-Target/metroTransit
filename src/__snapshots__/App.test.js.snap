// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h2
      class="page-title"
    >
      Real-time Departures
    </h2>
    <div
      class="select-box-container"
    >
      <select
        class="select-box"
      >
        <option>
          Select route
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
